rcl/src/rcl/type_version_hash.c

```diff
@@ -0,0 +1,257 @@
+      FAILFAST(
+        emit_individual_type_description(
+          emitter, &type_description->referenced_type_descriptions.data[i]));
+    }
+    FAILFAST(yaml_sequence_end_event_initialize(&event));
+    EMIT();
+  }
+
+  FAILFAST(yaml_mapping_end_event_initialize(&event));
+  EMIT();
+  return 1;
+}
+
+rcl_ret_t
+rcl_type_description_to_hashable_json(
+  const type_description_interfaces__msg__TypeDescription * type_description,
+  rcutils_char_array_t * output_repr)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_description, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(output_repr, RCL_RET_INVALID_ARGUMENT);
+
+  yaml_emitter_t emitter;
+  yaml_event_t event;
+
+  if (!yaml_emitter_initialize(&emitter)) {goto error;}
+  // Disable line breaks based on line length
+  yaml_emitter_set_width(&emitter, -1);
+  // Circumvent EOF line break by providing invalid break style
+  yaml_emitter_set_break(&emitter, -1);
+
+  yaml_emitter_set_output(&emitter, yaml_write_handler, output_repr);
+
+  if (
+    !yaml_stream_start_event_initialize(&event, YAML_UTF8_ENCODING) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+  if (
+    !yaml_document_start_event_initialize(&event, NULL, NULL, NULL, 1) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+
+  if (!emit_type_description(&emitter, type_description)) {goto error;}
+
+  if (
+    !yaml_document_end_event_initialize(&event, 1) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+  if (
+    !yaml_stream_end_event_initialize(&event) ||
+    !yaml_emitter_emit(&emitter, &event)) {goto error;}
+
+  yaml_emitter_delete(&emitter);
+  // Remove null-terminator character which should not be part of hashed content
+  output_repr->buffer_length--;
+  return RCL_RET_OK;
+
+error:
+  yaml_emitter_delete(&emitter);
+  return RCL_RET_ERROR;
+}
+
+rcl_ret_t
+rcl_calculate_type_version_hash(
+  const type_description_interfaces__msg__TypeDescription * type_description,
+  uint8_t output_hash[RCUTILS_SHA256_BLOCK_SIZE])
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(type_description, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(output_hash, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t result = RCL_RET_OK;
+  rcutils_char_array_t msg_repr = rcutils_get_zero_initialized_char_array();
+  msg_repr.allocator = rcl_get_default_allocator();
+
+  result = rcl_type_description_to_hashable_json(type_description, &msg_repr);
+  if (result != RCL_RET_OK) {
+    goto error;
+  }
+
+  rcutils_sha256_ctx_t sha_ctx;
+  rcutils_sha256_init(&sha_ctx);
+  // Last item in char_array is null terminator, which should not be hashed.
+  rcutils_sha256_update(&sha_ctx, (const uint8_t *)msg_repr.buffer, msg_repr.buffer_length);
+  rcutils_sha256_final(&sha_ctx, output_hash);
+
+  result = rcutils_char_array_fini(&msg_repr);
+  return result;
+
+error:
+  result = rcutils_char_array_fini(&msg_repr);
+  return RCL_RET_ERROR;
+}
```

Comment:
This can probably be simplified to:

```suggestion
  if (result == RCL_RET_OK) {
    rcutils_sha256_ctx_t sha_ctx;
    rcutils_sha256_init(&sha_ctx);
    // Last item in char_array is null terminator, which should not be hashed.
    rcutils_sha256_update(&sha_ctx, (const uint8_t *)msg_repr.buffer, msg_repr.buffer_length);
    rcutils_sha256_final(&sha_ctx, output_hash);
  }
  result = rcutils_char_array_fini(&msg_repr);
  return result;
}
```